### PR TITLE
Preview version of mgmt. profile package for Azure Stack

### PR DESCRIPTION
**Pull request by tadelesh**
Partially resolve https://github.com/Azure/azure-sdk-for-go/issues/16830.
- Add some packages to generate mgmt. plane config and SDK for profile.
- First version of profile config `2020-09-01-hybrid` package which only contains mgmt. plane modules.

### CODE REVIEW

File: sdk/profiles/hybrid20200901/go.mod
Comment (jhendrixMSFT) [inline sdk/profiles/hybrid20200901/go.mod:3]: This should be `go 1.18`

**Comment by jhendrixMSFT** [comment]
Few questions.

- Does this include full generated code, i.e. no removal of resource types?
- When will data-plane parts (i.e. key vault) be added?

**Comment by tadelesh** [comment]
> Few questions.
> 
> * Does this include full generated code, i.e. no removal of resource types?
> * When will data-plane parts (i.e. key vault) be added?

* Does this include full generated code, i.e. no removal of resource types?
Yes. I found it's hard to pair the config with operation/operation group. For example, which operation belong to `storageAccounts/blobServices` in the [spec](https://raw.githubusercontent.com/Azure/azure-rest-api-specs/master/specification/storage/resource-manager/Microsoft.Storage/stable/2019-06-01/storage.json) according to this [config](https://github.com/Azure/azure-rest-api-specs/blob/80065490402157d0df0dd37ab347c651b22eb576/profile/2020-09-01-hybrid.json#L462-L463). If I could figure out how to pair them, I think I could find some way to remove useless resource types.

* When will data-plane parts (i.e. key vault) be added?
Since we will announce the deprecation for track1 publicly later this September, I think we need to finish the work before that. According to the [issue](https://github.com/Azure/azure-sdk-for-go/issues/16830), it seems we don't have a conclusion for how we add the data plane modules. Is there any update for this?

File: sdk/profiles/hybrid20200901/resourcemanager/authorization/armauthorization/constants.go
Comment (jhendrixMSFT) [inline sdk/profiles/hybrid20200901/resourcemanager/authorization/armauthorization/constants.go:14]: I don't think this is correct. The module name and version should be declared at the module level, not per package. Will need codegen support.

Comment (tadelesh) [reply sdk/profiles/hybrid20200901/resourcemanager/authorization/armauthorization/constants.go:14]: For profile, we only have one module for all the RPs. If we only declare the model info in the root level, then telemetry data will not include any sub-package info. It that OK or we have other guideline for sub-package telemetry data format?

Comment (tadelesh) [reply sdk/profiles/hybrid20200901/resourcemanager/authorization/armauthorization/constants.go]: @jhendrixMSFT I have removed all the `moduleName` and `moduleVersion` from the sub-packages and introduce a root replacement. The side effect is that we need to export these two variables. Do you think it is OK?

File: sdk/profiles/hybrid20200901/constants.go
Comment (jhendrixMSFT) [inline sdk/profiles/hybrid20200901/constants.go:13]: This can go under an internal path to avoid exporting it.
```
sdk/profiles/hybrid20200901/internal
```

**Comment by tadelesh** [comment]
I've moved tooling change out of this PR and currently this PR should be a clear beta version for profile `hybrid20200901`, but only support mgmt. plane.

File: sdk/profiles/hybrid20200901/internal/constants.go
Comment (jhendrixMSFT) [inline sdk/profiles/hybrid20200901/internal/constants.go:15]: Missing new-line is causing the analysis failure.

File: sdk/profiles/hybrid20200901/internal/constants.go
Comment (jhendrixMSFT) [inline sdk/profiles/hybrid20200901/internal/constants.go:10]: Package name doesn't match directory name.

File: sdk/profiles/hybrid20200901/definition.json
Comment (jhendrixMSFT) [inline sdk/profiles/hybrid20200901/definition.json:1]: This info looks redundant with what's in the autorest.md files.  Can it be removed?

File: sdk/profiles/hybrid20200901/build.go
Comment (jhendrixMSFT) [inline sdk/profiles/hybrid20200901/build.go:5]: This will regenerate the entire profile right?

File: sdk/profiles/hybrid20200901/resourcemanager/authorization/armauthorization/autorest.md
Comment (jhendrixMSFT) [inline sdk/profiles/hybrid20200901/resourcemanager/authorization/armauthorization/autorest.md:11]: This, and `module-name`, should be removed as the sub packages aren't modules.

File: sdk/profiles/hybrid20200901/definition.json
Comment (tadelesh) [reply sdk/profiles/hybrid20200901/definition.json:1]: This config is used to generate autorest.md files for all the profile packages.

File: sdk/profiles/hybrid20200901/resourcemanager/authorization/armauthorization/autorest.md
Comment (tadelesh) [reply sdk/profiles/hybrid20200901/resourcemanager/authorization/armauthorization/autorest.md:11]: For current codegen, without module-version configured, arm package could not be generated successfully.

**Comment by jhendrixMSFT** [comment]
I see that the root `build.go` file was removed.  While you don't have to fix it now, we really should ensure that all modules can be built via `go generate`.

File: sdk/profiles/hybrid20200901/resourcemanager/authorization/armauthorization/autorest.md
Comment (jhendrixMSFT) [reply sdk/profiles/hybrid20200901/resourcemanager/authorization/armauthorization/autorest.md:11]: What was the error?  Is this somethin we need to fix in the code generator?

Comment (tadelesh) [reply sdk/profiles/hybrid20200901/resourcemanager/authorization/armauthorization/autorest.md:11]: Current codegen will always generate `moduleVersion` and `moduleName` constants ([code](https://github.com/Azure/autorest.go/blob/2fe13f887fd5a7def9a77590d294545239ec8433/src/generator/constants.ts#L34). And these two constants will be used when build pipeline ([code](https://github.com/Azure/autorest.go/blob/2fe13f887fd5a7def9a77590d294545239ec8433/src/generator/operations.ts#L214)). If we need to change codegen, we have to decide when it is a sub-package, what we should do for the pipeline building.

**Comment by tadelesh** [comment]
> I see that the root `build.go` file was removed. While you don't have to fix it now, we really should ensure that all modules can be built via `go generate`.

OK. I think if we decide to use profile, I'll do the fix.

File: sdk/profiles/hybrid20200901/resourcemanager/authorization/armauthorization/autorest.md
Comment (jhendrixMSFT) [reply sdk/profiles/hybrid20200901/resourcemanager/authorization/armauthorization/autorest.md:11]: Got it, thanks for clarifying.

Comment (tadelesh) [reply sdk/profiles/hybrid20200901/resourcemanager/authorization/armauthorization/autorest.md:11]: @jhendrixMSFT Regarding the solution you and Jeff propose, do we need to change code generator to support generate the sub package, or just like this PR, do some manual work or maybe could use directive to change the code?

Comment (jhendrixMSFT) [reply sdk/profiles/hybrid20200901/resourcemanager/authorization/armauthorization/autorest.md:11]: I suspect the solution will be a combination of small changes to the code generator and some autorest transforms.

**Comment by tadelesh** [comment]
@jhendrixMSFT @JeffreyRichter @RickWinter I've refined the PR with the new module path: `github.com/Azure/azure-sdk-for-go/profile/v20200928`. One thing I'm not sure is the `v20200928` is the pattern for Go module semantic version, will it be indexed by `pkg.go.dev` as profile's version `v20200928`?

**Comment by tadelesh** [comment]
@benbp It seems some scripts could not deal with our new module path, could you help?

**Comment by JeffreyRichter** [comment]
What is there to review really? If it appears in the repo correctly, then it's right.

**Comment by benbp** [comment]
@tadelesh sorry I'm missing context on why this isn't in the `sdk` folder? The `sdk/*` convention is a pretty hard requirement for almost all of our engineering system tools. If an exception is needed, we'll need to have a more in-depth conversation about what's possible and how much work it will involve.

File: profile/v20200901/ci.yml
Comment (benbp) [inline profile/v20200901/ci.yml:28]: How frequently will new profiles be added, that is to say, do we need to think about a large amount of pipelines getting created that differ by profile date/version?

Comment (tadelesh) [reply profile/v20200901/ci.yml:28]: I think there won't be too much new profile, maybe one per year or even longer interval, it is only for Azure Stack. Thus, I could bear with little manual work to let pipeline work.

**Comment by tadelesh** [comment]
> @tadelesh sorry I'm missing context on why this isn't in the `sdk` folder? The `sdk/*` convention is a pretty hard requirement for almost all of our engineering system tools. If an exception is needed, we'll need to have a more in-depth conversation about what's possible and how much work it will involve.

We have a discussion about how to deal with profile for our Go track2 SDK recently. As we do not want to miss-lead customer to wrongly use profile module, we are thinking about put them in `profile` parallelly to `sdk`.

**Comment by tadelesh** [comment]
@jhendrixMSFT I've updated the PR, could you help to approve?

**Comment by benbp** [comment]
Ok that yaml is good with me @tadelesh

**Comment by tadelesh** [comment]
Then it seems everything is good. @jhendrixMSFT do you have any further concern? After your approval, I will merge and release the first beta.

File: profile/v20200901/resourcemanager/appservice/armappservice/client_factory.go
Comment (jhendrixMSFT) [inline profile/v20200901/resourcemanager/appservice/armappservice/client_factory.go:32]: This is going to be ambiguous for all client factories.

File: profile/v20200901/resourcemanager/appservice/armappservice/certificates_client.go
Comment (jhendrixMSFT) [inline profile/v20200901/resourcemanager/appservice/armappservice/certificates_client.go:37]: This might also be ambiguous for some clients.

File: profile/v20200901/resourcemanager/appservice/armappservice/client_factory.go
Comment (tadelesh) [reply profile/v20200901/resourcemanager/appservice/armappservice/client_factory.go:32]: Oh, yes. It will cause distributed tracing ambiguous. I added a package prefix to it (`internal.ModuleName+"/armappservice.ClientFactory"`). Is that OK?

Comment (jhendrixMSFT) [reply profile/v20200901/resourcemanager/appservice/armappservice/client_factory.go]: This will show up in the telemetry data as `v20200901/armappservice.ClientFactory`.  How much do we care about being able to distinguish the profile version of this client with the non-profile version?

CC @JeffreyRichter

Comment (tadelesh) [reply profile/v20200901/resourcemanager/appservice/armappservice/client_factory.go]: The complete telemetry will be `azsdk-go-v20200901/armappservice/v0.1.0`. A little bit ugly, but I think it is useful when analyzing user data.

Comment (JeffreyRichter) [reply profile/v20200901/resourcemanager/appservice/armappservice/client_factory.go]: I'm OK with this too.

**Review by jhendrixMSFT — APPROVED** [review]

**Pull request merged**
